rp2040-hal/src/timer.rs

```diff
@@ -106,6 +112,63 @@ impl Timer {
 // safety: all write operations are synchronised and all reads are atomic
 unsafe impl Sync for Timer {}
 
+/// [`Timer`] based Delay
+pub struct Delay(());
+impl Delay {
+    fn delay_from_u32(&self, us: u32) {
+        // safety: only reads are made
+        let timer = unsafe { &*pac::TIMER::PTR };
+        let eta = timer.timelr.read().bits().wrapping_add(us);
+        while eta >= timer.timelr.read().bits() {
+            // use nop instead of wfe because there is no event configured to wake us up.
+            cortex_m::asm::nop();
+        }
+    }
+
+    fn delay_from_u64(&self, us: u64) {
+        // safety: only reads are made
+        let timer = unsafe { &*pac::TIMER::PTR };
+        // wrapping_add isn't strictly necessary here, there's no way we'll reach the wrap point of
+        // that u64, right?
```

Comment:
It may not be useful, but `delay_us(u64::MAX)` will probably reach that wrap point.

Which makes me wonder: Is it completely unrealistic that some code might do that? Shouldn't that just block forever (for all practical matters), instead of returning immediately?

→ perhaps it would be better to use `saturating_add`?

Reply:
using the `wrapping_sub` method should avoid the issue all together :D